### ISSUE TITLE
Add a way to execute suggestions

### ISSUE THREAD

**Issue by emillon** [ocaml/dune#1404]
Dune can help detect missing libraries, but at the moment it's very copy & pasty:

```
 % dune build tt.exe
File "dune", line 3, characters 12-18:
3 |  (libraries zxcvbn)
                ^^^^^^
Error: Library "zxcvbn" not found.
Hint: try: dune external-lib-deps --missing tt.exe
 % dune external-lib-deps --missing tt.exe
Done: 0/0 (jobs: 0)Error: The following libraries are missing in the default context:
- zxcvbn
Hint: try: opam install zxcvbn
 /tmp/ex % opam install zxcvbn
The following actions will be performed:
  ∗ install zxcvbn 2.3+1

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
Processing  1/1: [zxcvbn.2.3+1: dl]
```

Something more integrated would be welcome, e.g.

```
 % dune build tt.exe
File "dune", line 3, characters 12-18:
3 |  (libraries zxcvbn)
                ^^^^^^
Error: Library "zxcvbn" not found.
It looks like you are missing the following external libraries:
- zxcvbn
Shall I try to install them? [Y/n] y
The following actions will be performed:
  ∗ install zxcvbn 2.3+1

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
Processing  1/1: [zxcvbn.2.3+1: dl]
```

cc @linse @samoht

**Comment by diml** [comment]
That would be great!

**Comment by rgrinberg** [comment]
Sounds good, but rather than prompting, which is very annoying and can cause issues in automated environments, can we reuse the promotion mechanism for this?

**Comment by samoht** [comment]
@rgrinberg for automated environment, I guess you can add `-y` and/or `DUNEYES=true` :-)

**Comment by rgrinberg** [comment]
More like `-n` and `DUNEYES=false` :) But anyways, we'd have to agree that there would be an option to disable all prompts and do not make any changes.

**Comment by andreypopp** [comment]
Executing opam by dune automatically would be very annoying in non opam environments.

**Comment by emillon** [comment]
Yes, this will require good detection of interactive environment. We don't want to break scripts or non-opam installs, but at the same time having to copy & paste two commands is not great. Just making that a single step would be useful I think.

**Comment by rgrinberg** [comment]
Do you mean promotion would not be a great experience?

Btw, another feature which is more valuable and would largely supercede this would be LSP support for dune files. The LSP protocol covers auto corrections like this, so it would cover corrections like this and more.

I suppose the code that would be implemented here would be reusable for both prompts and LSP, so I'm cautiously in favor this. But let's definitely keep the prompts off by default.

**Comment by emillon** [comment]
I don't think we're talking about the same problem. I'm referring to the case where you `git clone`d a random repository, you want to build the project, but for some reason you don't have all the dependencies installed.

In that case, the `dune` file is fine, and it does not need to be fixed. But the user has to install some packages, which dune knows about. So, it could be better in indicating (or doing) what the user has to do to resolve the issue.

**Comment by rgrinberg** [comment]
Oops, you're right that I miunderstood. I still see the problem that is being described here as artificial. If I clone a project, the first thing that I run is `$ opam install --deps-only .` or `$ esy install`. This is both more accurate than w/e suggestion dune can make and it doesn't prefer one package manager over another.

**Comment by samoht** [comment]
>  I still see the problem that is being described here as artificial

It's not artificial -- it happens regularly to me when I clone repos in you "monorepo" dev tree, in that case I cannot use opam as I would need to pin the subdirectories first.

**Comment by rgrinberg** [comment]
`$ opam install --deps-only .` doesn't require pinning. Does it not recurse into the subdirs?

It might not be entirely artifical, but dune can't really know for sure which opam packages you need. If you already have opam files, why are you against solving this problem at the opam level? dune isn't a package manager and opam already has a solution for this. I don't see too many benefits in re-implementing this in dune.

**Comment by samoht** [comment]
I don't think opam will recurse in the subdirectories.

Dune already have the information as it displays it as a hint. Usually it works very well (and it will work more and more given that nowadays people are now behaving nicely and are choosing the same name for opam and ocamlfind packages).

Anyway, that's an issue that I have very regularly (and a few people hit that as well in the hack retreat)  so I think it's worth improving the UX a bit.

**Comment by rgrinberg** [comment]
I agree there's a problem with the UI, it's just a problem with opam's UI and not dune's. Opam 2.0 already has a good workflow for installing whatever packages are missing, and esy improves on it further still.

Also, my experience contradicts your experience at the retreat. Most users that I've seen usually look for an install step to install the dependencies. Only after this step, they will try to build the package. This is how npm, cargo, maven, and a few other package managers work. I have not seen any prompt based installation anywhere else on the other hand.

**Comment by samoht** [comment]
So are you saying that `opam` should scan recursively for all opam packages in the tree and `opam install . --deps` should install everything that is needed?

**Comment by rgrinberg** [comment]
That seems sensible. That's how it already works if there are several opam files in a directory. I assume this was added because installing the dependencies for every individual packages was getting tedious. If people are switching to a directory layout where the opam package files are nested, it would be helpful for `$ opam install` to accommodate this.

**Comment by rgrinberg** [comment]
>  If that's the case, then dune should be changed to give the right hints instead of asking the user to copy/paste some hints twice.

Thinking about this again, dune's hint could be improved by telling users to run `$ opam install --deps-only <all the dirs that contain .opam files>`.

But it's a still deficiency in opam that we need this and we don't just have one command to install the dependencies

**Comment by diml** [comment]
In general, it's nice if we can reduce the boilerplate to a minimum. If Dune detects that the user is using opam and is confident that the next step the user should take is run `opam blah`, then I believe dune should make it as easy as possible for the user to go ahead running `opam blah`.

**Comment by rgrinberg** [comment]
How would you accurately detect that the user is using opam? I don't use opam all that often myself, and there are now two alternatives to it: duniverse and esy. Both of which don't suffer from the problem of installing dependencies when building a cloned package. Dune has stayed package manager agnostic, it's a shame to break that. Especially in a situation like this where we're basically just working around opam's deficiencies. This suggest would be fairly useless for users of esy/duniverse.

**Comment by emillon** [comment]
I don't think we're trying to break anything here: just following the instructions today will suggest to use `opam`. This instruction is already wrong for cases where `duniverse` or `esy` is used.

IMO there are two independent things we can do in dune to ease this:
- somehow detect a more precise better command (I don't know how to do this, and I agree that this looks hard!)
- shorten the distance between the error message and the resolution (this issue)

**Comment by diml** [comment]
To detect the use of opam, we can look at the environment variable `OPAM_SWITCH_PREFIX` for instance. We already do that in `src/context.ml`.

I agree that it's nice for Dune to be package manager agnostic, but at the same time it's nice to provide enhanced integration when we can and when it improves the user experience. Currently, there are a lot of developers using ocaml+opam+dune for their day-to-day development and such a feature would make things easier for them. So if someone wants to work on such a feature, I don't think that this is a bad thing.

Additionally, a general mechanism for asking the user for feedback could probably be useful in other cases, so it's nice to have this possibility in dune. In particular, I would expect that the opam specific part of such a feature would be relatively small compared to the general mechanism.

**Comment by rgrinberg** [comment]
> somehow detect a more precise better command (I don't know how to do this, and I agree that this looks hard!)

What about my suggestion of using `$ opam install . <dirs>`. If passing <dirs> is tedious, then opam should be improved.

> To detect the use of opam, we can look at the environment variable OPAM_SWITCH_PREFIX for instance. We already do that in src/context.ml.

I don't think this an accurate indicator. I have this variable turned set, but I certainly don't use opam for the majority of my projects.

I'm not against a more integrated workflow for dune+opam users by the way. I agree that such a thing is desirable, even if I think it's misguided in this particular case. I'm simply opposed in making an opam centric workflow the default. Perhaps we can find a better indicator for such projects. How about we check for a project local switch?

**Comment by diml** [comment]
Personally, I don't know the new behaviors enough to comment on `opam install . <dirs>` vs `opam install blah`.

> I'm not against a more integrated workflow for dune+opam users by the way. I agree that such a thing is desirable, even if I think it's misguided in this particular case. I'm simply opposed in making an opam centric workflow the default. Perhaps we can find a better indicator for such projects. How about we check for a project local switch?

Yh, I agree that we shouldn't be opam centric given that dune is now used in combination with other tools. Regarding local switches, I have no opinion as I don't know if they are used a lot. Personally, I tried them but since you have to recompile a new compiler every time, I find them too painful to use.

if there is no good way to detect the use of opam, that's indeed a problem :/
